module/blobs/blob_io.go

```diff
@@ -0,0 +1,305 @@
+package blobs
+
+import (
+	"errors"
+	"io"
+	"sync"
+)
+
+const defaultInitialBufCapacity = 1 << 17 // 128 KiB
+
+var ErrClosedBlobChannel = errors.New("send/receive on closed blob channel")
+
+// blobChannel represents a channel of blobs which can be closed without causing a panic
+type blobChannel struct {
+	blobs chan Blob
+	once  sync.Once
+	done  chan struct{}
+}
+
+// Send sends a blob to the blob channel. It returns ErrClosedBlobChannel if the channel is closed.
+func (bc *blobChannel) Send(blob Blob) error {
+	select {
+	case <-bc.done:
+		return ErrClosedBlobChannel
+	default:
+	}
+
+	select {
+	case bc.blobs <- blob:
+		return nil
+	case <-bc.done:
+		return ErrClosedBlobChannel
+	}
+}
+
+// Receives receives a blob from the blob channel. It returns ErrClosedBlobChannel if the channel is closed.
+func (bc *blobChannel) Receive() (Blob, error) {
+	select {
+	case <-bc.done:
+		return nil, ErrClosedBlobChannel
+	default:
+	}
+
+	select {
+	case blob := <-bc.blobs:
+		return blob, nil
+	case <-bc.done:
+		return nil, ErrClosedBlobChannel
+	}
+}
+
+// Close closes the blob channel. The returned error is always nil.
+func (bc *blobChannel) Close() error {
+	bc.once.Do(func() { close(bc.done) })
+	return nil
+}
```

Comment:
See the go source code here: https://cs.opensource.google/go/go/+/refs/tags/go1.17.3:src/io/pipe.go;l=75-82

I am definitely very open to discussing alternative code, but just want to point out that this is a pattern that exists in the Go source code, so it can't be that much of an antipattern.

Reply:
I didn't realize that you were implementing a pipe rather than just wrapping channels, my bad. I'm not sure I understand, in this case, the advantage/reason for using a pipe system over a channel? I guess there is some complexity I'm not seeing here that justifies it, so I might be missing something 😄

Reply:
@Ullaakut  Yeah, to be frank the main reason for introducing this is because we need some mechanism for the receiver to close the pipe, signalling the sender to exit. And as you said, with channels the convention is that the sender always closes the channel.

In our case, the receiver is either doing some sort of validation (ie deserializing the data) or writing the received data to some external service (blob service), during which it may encounter errors. If it does, I wanted a simple way for it to signal the sender to stop sending data. Also, the receiver may be the one who has access to the context, which may be canceled.

If you have a suggestion on how to better handle this usecase I am very open to it btw. I thought about other ways but in the end this one seemed the cleanest to me.